A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,668 @@
+# gRPC xDS traffic splitting and routing
+----
+* Author(s): Menghan Li
+* Approver: markdroth
+* Status: Draft
+* Implemented in: (in progress in C-core, Java, and Go)
+* Last updated: 2020-05-20
+* Discussion at: https://groups.google.com/g/grpc-io/c/2Ot9HA3Rjeg
+
+
+
+
+# Background
+
+Traffic splitting is a feature where RPCs are sent to different clusters with
+a specified distribution. RPC routing is a feature where RPCs are sent to
+different clusters based on different criteria (e.g. path matching, header
+matching). In xDS, routing configurations are specified in RDS responses,
+including different ways to match a route, and different actions to perform
+to split the traffic.
+
+This doc covers how gRPC parses RDS routing related fields, and how to route
+the RPCs.
+
+It's based on the resolver/LB architecture and the xDS client behavior in the
+[A27: xDS-Based Global Load
+Balancing](https://github.com/grpc/proposal/blob/xds/A27-xds-global-load-balancing.md).
+
+
+# Overview
+
+To support traffic splitting and routing, the xDS resolver will be extended
+to handle the routing fields in RDS responses, and convert them to service
+config.
+
+A new top level LB policy (`xds_routing_experimental`) will be added to do
+RPC routing. It will pick the action to perform for the matched RPCs.
+
+The action can be either to route to a cluster, or to do a weighted cluster
```

Comment:
Do you want to cover what will happen if an unsupported action is specified in the `RouteConfiguration`?

Reply:
In general, we've decided to stick with the "ignore unsupported fields" approach.  This is what happens when new fields are added to a proto API, so it's something users need to be aware of anyway.

We've had quite a bit of debate over this, and I'm not entirely happy with the outcome, because I do think that there are cases where this can cause surprising behavior.  But we've tried to mitigate this by including as many of the matchers as we can in this design, since ignoring a matcher can cause much more surprising behavior than ignoring an action.  Ignoring an action generally just means that a feature is not yet available -- e.g., if retries are configured and we don't retry, it's probably not the end of the world.

Reply:
Fair, but it might be reasonable to state this explicitly.

Reply:
Good point.  Menghan, please add a note somewhere indicating that this is our general approach to unsupported fields in the xDS API.

Reply:
In the "Response validation" section, [line 401](https://github.com/grpc/proposal/pull/178/files#diff-86f80616477d2453f4ea068d913c9b0aR401), there's
> The fields not covered by this section will all be ignored.

It seems to be the right section for this, because it's where we talk about each field in the response.

Let me know if you think there's a better place for it.

Reply:
LGTM, thanks!